monai/data/utils.py

```diff
@@ -315,63 +315,38 @@ def decollate_batch(data: Union[dict, list, torch.Tensor], batch_size: Optional[
 
     Args:
-        data: data to be de-collated.
-        batch_size: number of batches in data. If `None` is passed, try to figure out batch size.
-    """
-
-    def torch_to_single(d: torch.Tensor):
-        """If input is a torch.Tensor with only 1 element, return just the element."""
-        return d if d.numel() > 1 else d.item()
-
-    def decollate(data: Any, idx: int):
-        """Recursively de-collate."""
-        if isinstance(data, dict):
-            return {k: decollate(v, idx) for k, v in data.items()}
-        if isinstance(data, torch.Tensor):
-            out = data[idx]
-            return torch_to_single(out)
-        if isinstance(data, list):
-            if len(data) == 0:
-                return data
-            if isinstance(data[0], torch.Tensor):
-                return [torch_to_single(d[idx]) for d in data]
-            if issequenceiterable(data[0]):
-                return [decollate(d, idx) for d in data]
-            return data[idx]
-        raise TypeError(f"Not sure how to de-collate type: {type(data)}")
-
-    def _detect_batch_size(batch_data):
-        for v in batch_data:
-            if isinstance(v, torch.Tensor):
-                return v.shape[0]
-        warnings.warn("batch_data is not a sequence of tensors in decollate, use `len(batch_data[0])` directly.")
-        return len(batch_data[0])
-
-    result: List[Any]
-    if isinstance(data, dict):
-        batch_size = _detect_batch_size(batch_data=data.values()) if batch_size is None else batch_size
-        result = [{key: decollate(data[key], idx) for key in data.keys()} for idx in range(batch_size)]
-    elif isinstance(data, list):
-        batch_size = _detect_batch_size(batch_data=data) if batch_size is None else batch_size
-        result = [[decollate(d, idx) for d in data] for idx in range(batch_size)]
-    elif isinstance(data, torch.Tensor):
-        batch_size = data.shape[0]
-        result = [data[idx] for idx in range(batch_size)]
-    else:
-        raise NotImplementedError("Only currently implemented for dictionary, list or Tensor data.")
-
-    return result
+        batch: data to be de-collated.
+        detach: whether to detach the tensors. Scalars tensors will be detached into number types
+            instead of torch tensors.
+    """
+    if isinstance(batch, torch.Tensor):
+        if detach:
+            batch = batch.detach()
+        if batch.ndim == 0 and detach:
+            return batch.item()
+        out_list = torch.unbind(batch, dim=0)
+        if out_list[0].ndim == 0 and detach:
+            return [t.item() for t in out_list]
+        return list(out_list)
+    if isinstance(batch, Mapping):
+        _dict_list = {key: decollate_batch(batch[key]) for key in batch}
+        return [dict(zip(_dict_list, item)) for item in zip(*_dict_list.values())]
+    if isinstance(batch, Iterable):
+        item_0 = first(batch)
+        if isinstance(item_0, Iterable) and not isinstance(item_0, (str, bytes)):
```

Comment:
Just to confirm: so here you suppose every item in the `batch` should be Iterable?

Reply:
Because I am thinking about how to decollate `engine.state.output` which contains `"loss"： XXX`, which is a scalar number without batch dim?

Thanks.

Reply:
could you make a test case to show the usage? I'll update the code, currently it wouldn't decollate

Reply:
I mean for example:
```py
engine.state.output = {
    "image": torch.tensor([[[1, 2]], [[3, 4]]]),
    "label": torch.tensor([[[5, 6]], [[7, 8]]]),
    "pred": torch.tensor([[[9, 10]], [[11, 12]]]),
    "loss": 0.56,
}
```
I think your current code can handle `image`, `label`, `pred` then make a list of dictionaries, but how do you handle the `loss` in this case? Just copy to every dictionary in the decollated list? like:
```
[
    {"image": torch.tensor([[1, 2]]), "label": torch.tensor([[5, 6]]), "pred": torch.tensor([[9, 10]]), "loss": 0.56},
    {"image": torch.tensor([[3, 4]]), "label": torch.tensor([[7, 8]]), "pred": torch.tensor([[11, 12]]), "loss": 0.56},
]
```
How about if `engine.state.output` is a list and the **first item is loss**?
```py
engine.state.output = [0.56, torch.tensor([[[1, 2]], [[3, 4]]]), torch.tensor([[[5, 6]], [[7, 8]]]), torch.tensor([[[9, 10]], [[11, 12]]])]
```

Thanks.

Reply:
ok, I feel these are out of the scope of this utility, this function will only do the 'inverse' according to pytorch's 'default_collate'
mainly:
```py
>>> import torch
>>> from torch.utils.data._utils.collate import default_collate
>>> from monai.data.utils import decollate_batch
>>> default_collate([[0, 0]])
[tensor([0]), tensor([0])]
>>> decollate_batch([torch.tensor([0]), torch.tensor([0])])
[[0, 0]]
>>> 
>>> default_collate([0])
tensor([0])
>>> decollate_batch(torch.tensor([0]))
[0]

```

Reply:
Sure, this PR is good enough for the current situation, let's discuss the scalar case later in my PR: https://github.com/Project-MONAI/MONAI/pull/2315.

Thanks.